# hypedsearch
## **hy**brid **pe**pti**d**e **search**ing tool

For more information about hypedsearch, installation and usage instructions, please see the documentation [found here](https://hypedsearch.readthedocs.io/en/latest/). 

---

Background
T1D is an autoimmune disesase where the body's immune system attacks and destroy's the panceas's ability to produce insulin<sup>1</sup>.  


Dr. Delong from CU Anchutez has identified a possible cause of Type 1 Diabetes (T1D) - hybryd insuline peptdes (reference needed). 

It appears that in Type 1 Diabetics, the     


`hypedsearch` is a tool for identifying both hybrid and non-hybrid proteins from mass spectrometry data. `hypedsearch` takes `mzML` and `fasta` files as inputs and outputs a dataframe with a quality score.

`hypedsearch` identifies sequences by using a method called "k-mer extension".  Proteins are composed of a linear strand of amino acids 

If you're not familiar, a "k-mer" is a k long string, or in this case, k long sequence of amino acids. The process, at a high level, works like this:


## References
##Diabetes
1) https://en.wikipedia.org/wiki/Type_1_diabetes


### pyteomics
*Goloborodko, A.A.; Levitsky, L.I.; Ivanov, M.V.; and Gorshkov, M.V. (2013) “Pyteomics - a Python Framework for Exploratory Data Analysis and Rapid Software Prototyping in Proteomics”, Journal of The American Society for Mass Spectrometry, 24(2), 301–304. DOI: 10.1007/s13361-012-0516-6

Levitsky, L.I.; Klein, J.; Ivanov, M.V.; and Gorshkov, M.V. (2018) “Pyteomics 4.0: five years of development of a Python proteomics framework”, Journal of Proteome Research. DOI: 10.1021/acs.jproteome.8b00717
